src/OAuth1/AbstractProvider.php

```diff
@@ -137,54 +131,98 @@ public function parseToken($body)
         return new Token($token['oauth_token'], $token['oauth_token_secret']);
     }
 
+    /**
+     * @param string $method
+     * @param string $uri
+     * @param array $query
+     * @return string
+     */
+    public function getSignatureBaseString(string $method, string $uri, array $query): string
+    {
+        $parts = [
+            $method,
+            $uri,
+            Util::buildHttpQuery(
+                $query
+            )
+        ];
+
+        $parts = Util::urlencodeRFC3986($parts);
+
+        return implode('&', $parts);
+    }
+
+    public function authorizationHeader(array $query)
+    {
+        $parameters = http_build_query(
+            $query,
+            '',
+            ', ',
+            PHP_QUERY_RFC3986
+        );
+
+        return "OAuth $parameters";
+    }
+
     /**
      * @param string $uri
      * @param string $method
-     * @param array $parameters
-     * @return \SocialConnect\Common\Http\Response
+     * @param array $query
+     * @param null $payload
+     * @param array $headers
+     * @return ResponseInterface
+     * @throws InvalidResponse
+     * @throws \Psr\Http\Client\ClientExceptionInterface
      */
-    public function oauthRequest($uri, $method = Client::GET, $parameters = [], $headers = [])
+    public function oauthRequest($uri, $method = 'GET', array $query = [], $payload = null, $headers = []): ResponseInterface
     {
         $headers = array_merge([
             'Accept' => 'application/json'
         ], $headers);
 
-        if ($method == Client::POST) {
+        if ($method === 'POST') {
             $headers['Content-Type'] = 'application/x-www-form-urlencoded';
         }
 
-        $parameters = array_merge(
-            [
-                'oauth_version' => '1.0',
-                'oauth_nonce' => md5(time() . mt_rand()),
-                'oauth_timestamp' => time(),
-                'oauth_consumer_key' => $this->consumer->getKey()
-            ],
-            $parameters
-        );
-
-        $request = new Request(
-            $uri,
-            $parameters,
-            $method,
-            $headers
-        );
+        $query['oauth_version'] = '1.0';
+        $query['oauth_nonce'] = md5(time() . mt_rand());
```

Comment:
Ideally the nonce should be cryptographically secure. So using something like `bin2hex(random_bytes(16))` would be better.

Reply:
Yeap, it's better, `OAuth2` already use it

```php
    /**
     * 16 bytes / 128 bit / 16 symbols / 32 symbols in hex
     */
    const STATE_BYTES = 16;

    /**
     * @return string
     * @throws \Exception
     */
    protected function generateState(): string
    {
        return bin2hex(random_bytes(self::STATE_BYTES));
    }
```

Fixed! :)